pkg/cloud/cloud.go

```diff
@@ -2438,28 +2457,141 @@ func getVolumeAttachmentsList(volume types.Volume) []string {
 }
 
 // Calculate actual IOPS for a volume and cap it at supported AWS limits.
-func capIOPS(volumeType string, requestedCapacityGiB int32, requestedIops int32, minTotalIOPS, maxTotalIOPS, maxIOPSPerGB int32, allowIncrease bool) int32 {
+func capIOPS(volumeType string, requestedCapacityGiB int32, requestedIops int32, iopsLimits iopsLimits, allowIncrease bool) int32 {
 	// If requestedIops is zero the user did not request a specific amount, and the default will be used instead
 	if requestedIops == 0 {
 		return 0
 	}
 
 	iops := requestedIops
 
-	if iops < minTotalIOPS {
+	if iops < iopsLimits.minIops {
 		if allowIncrease {
-			iops = minTotalIOPS
+			iops = iopsLimits.minIops
 			klog.V(5).InfoS("[Debug] Increased IOPS to the min supported limit", "volumeType", volumeType, "requestedCapacityGiB", requestedCapacityGiB, "limit", iops)
 		}
 	}
-	if iops > maxTotalIOPS {
-		iops = maxTotalIOPS
+	if iops > iopsLimits.maxIops {
+		iops = iopsLimits.maxIops
 		klog.V(5).InfoS("[Debug] Capped IOPS, volume at the max supported limit", "volumeType", volumeType, "requestedCapacityGiB", requestedCapacityGiB, "limit", iops)
 	}
-	maxIopsByCapacity := maxIOPSPerGB * requestedCapacityGiB
-	if iops > maxIopsByCapacity && maxIopsByCapacity >= minTotalIOPS {
+	maxIopsByCapacity := iopsLimits.maxIopsPerGb * requestedCapacityGiB
+	if iops > maxIopsByCapacity && maxIopsByCapacity >= iopsLimits.minIops {
 		iops = maxIopsByCapacity
-		klog.V(5).InfoS("[Debug] Capped IOPS for volume", "volumeType", volumeType, "requestedCapacityGiB", requestedCapacityGiB, "maxIOPSPerGB", maxIOPSPerGB, "limit", iops)
+		klog.V(5).InfoS("[Debug] Capped IOPS for volume", "volumeType", volumeType, "requestedCapacityGiB", requestedCapacityGiB, "maxIOPSPerGB", iopsLimits.maxIopsPerGb, "limit", iops)
 	}
 	return iops
 }
+
+// Gets IOPS limits for a specific volume type in a specific Zone and caches it. If the limits are cached, simply return limits.
+func (c *cloud) getVolumeLimits(ctx context.Context, volumeType string, azParams paramsForGetVolumeLimits) (iopsLimits iopsLimits, err error) {
+	az := azParams.AvailabilityZone
+	azId := azParams.AvailabilityZoneId
+	outpostArn := azParams.OutpostArn
+
+	cacheKey := fmt.Sprintf("%s|%s|%s|%s", volumeType, az, azId, outpostArn)
+	if c.latestIOPSLimits != nil {
+		if value, ok := c.latestIOPSLimits.Get(cacheKey); ok {
+			return *value, nil
+		}
+	}
+
+	dryRunRequestInput := &ec2.CreateVolumeInput{
+		VolumeType: types.VolumeType(volumeType),
+		Size:       aws.Int32(1),
+		Iops:       aws.Int32(math.MaxInt32),
+		DryRun:     aws.Bool(true),
+		// Required by IAM.
```

Comment:
Make this comment more clear - e.g. "Required by default EBS CSI Driver IAM policy"